linear_programming/simplex.py

```diff
@@ -0,0 +1,318 @@
+
+        # Delete the objective row of the first stage
+        tableau = np.delete(tableau, 0, axis=0)
+
+        self.n_stages = 1
+        self.n_rows -= 1
+        self.n_art_vars = 0
+        self.stop_iter = False
+        return tableau
+
+    def run_simplex(self) -> dict[Any, Any]:
+        """Operate on tableau until objective function cannot be
+        improved further.
+
+        # Standard linear program:
+        Max:  x1 +  x2
+        ST:   x1 + 3x2 <= 4
+             3x1 +  x2 <= 4
+        >>> Tableau(np.array([[-1,-1,0,0,0],[1,3,1,0,4],[3,1,0,1,4.]]),
+        ... 2).run_simplex()
+        {'P': 2.0, 'x1': 1.0, 'x2': 1.0}
+
+        # Optimal tableau input:
+        >>> Tableau(np.array([
+        ... [0, 0, 0.25, 0.25, 2],
+        ... [0, 1, 0.375, -0.125, 1],
+        ... [1, 0, -0.125, 0.375, 1]
+        ... ]), 2).run_simplex()
+        {'P': 2.0, 'x1': 1.0, 'x2': 1.0}
+
+        # Non-standard: >= constraints
+        Max: 2x1 + 3x2 +  x3
+        ST:   x1 +  x2 +  x3 <= 40
+             2x1 +  x2 -  x3 >= 10
+                 -  x2 +  x3 >= 10
+        >>> Tableau(np.array([
+        ... [2, 0, 0, 0, -1, -1, 0, 0, 20],
+        ... [-2, -3, -1, 0, 0, 0, 0, 0, 0],
+        ... [1, 1, 1, 1, 0, 0, 0, 0, 40],
+        ... [2, 1, -1, 0, -1, 0, 1, 0, 10],
+        ... [0, -1, 1, 0, 0, -1, 0, 1, 10.]
+        ... ]), 3).run_simplex()
+        {'P': 70.0, 'x1': 10.0, 'x2': 10.0, 'x3': 20.0}
+
+        # Non standard: minimisation and equalities
+        Min: x1 +  x2
+        ST: 2x1 +  x2 = 12
+            6x1 + 5x2 = 40
+        >>> Tableau(np.array([
+        ... [8, 6, 0, -1, 0, -1, 0, 0, 52],
+        ... [1, 1, 0, 0, 0, 0, 0, 0, 0],
+        ... [2, 1, 1, 0, 0, 0, 0, 0, 12],
+        ... [2, 1, 0, -1, 0, 0, 1, 0, 12],
+        ... [6, 5, 0, 0, 1, 0, 0, 0, 40],
+        ... [6, 5, 0, 0, 0, -1, 0, 1, 40.]
+        ... ]), 2).run_simplex()
+        {'P': 7.0, 'x1': 5.0, 'x2': 2.0}
+        """
+        # Stop simplex algorithm from cycling.
+        iter_num = 0
+
+        while iter_num < 100:
+            # Completion of each stage removes an objective. If both stages
+            # are complete, then no objectives are left
+            if not self.objectives:
+                self.col_titles = self.generate_col_titles(
+                    self.n_vars, self.n_slack, self.n_art_vars
+                )
+
+                # Find the values of each variable at optimal solution
+                return self.interpret_tableau(self.tableau, self.col_titles)
+
+            row_idx, col_idx = self.find_pivot(self.tableau)
+
+            # If there are no more negative values in objective row
+            if self.stop_iter:
+                # Delete artificial variable columns and rows. Update attributes
+                self.tableau = self.change_stage(self.tableau)
+
+                # Pivot again
+                continue
+
+            self.tableau = self.pivot(self.tableau, row_idx, col_idx)
+            iter_num += 1
```

Comment:
```suggestion
        for _ in range(100):
            # Completion of each stage removes an objective. If both stages
            # are complete, then no objectives are left
            if not self.objectives:
                self.col_titles = self.generate_col_titles(
                    self.n_vars, self.n_slack, self.n_art_vars
                )

                # Find the values of each variable at optimal solution
                return self.interpret_tableau(self.tableau, self.col_titles)

            row_idx, col_idx = self.find_pivot(self.tableau)

            # If there are no more negative values in objective row
            if self.stop_iter:
                # Delete artificial variable columns and rows. Update attributes
                self.tableau = self.change_stage(self.tableau)
            else:
                self.tableau = self.pivot(self.tableau, row_idx, col_idx)
```
For-loops are more efficient than while-loops in Python, and they're easier to read in this case, IMO. Also,  the value of `iter_num` is never actually used

Reply:
Yes, agreed and thank you